enhancements/dns/forward-dns-over-tls.md

```diff
@@ -0,0 +1,317 @@
+  
+  // tls contains the additional configuration options to use when Transport is set to "TLS".
+  TLS *DNSOverTLSConfig `json:"tls,omitempty"`
+}
+
+// DNSOverTLSConfig describes optional DNSTransportConfig fields that should be captured.
+type DNSOverTLSConfig struct {
+  // serverName is the upstream server to connect to when forwarding DNS queries.
+  ServerName string `json:"serverName"`
+  
+  // caBundle references a ConfigMap that must contain either a single
+  // CA Certificate or a CA Bundle (in the case of multiple upstreams signed
+  // by different CAs).
+  CABundle v1.ConfigMapNameReference `json:"caBundle,omitempty"`
+}
+```
+
+`UpstreamResolvers` Modifications:
+```go
+type UpstreamResolvers struct {
+  // <snip>
+  
+  // transportConfig is used to configure the transport type, server name, and optional custom CA or CA bundle to use
+  // when forwarding DNS requests to an upstream resolver.
+  //
+  // The default value is "" (empty) which results in a standard cleartext connection being used when forwarding DNS
+  // requests to an upstream resolver.
+  //
+  // +optional
+  TransportConfig DNSTransportConfig `json:"transportConfig,omitempty"`
+}
+```
+
+`ForwardPlugin` modifications:
+```go
+type ForwardPlugin struct {
+  // <snip>
+  
+  // transportConfig is used to configure the transport type, server name, and optional custom CA or CA bundle to use
+  // when forwarding DNS requests to an upstream resolver.
+  //
+  // The default value is "" (empty) which results in a standard cleartext connection being used when forwarding DNS
+  // requests to an upstream resolver.
+  //
+  // +optional
+  TransportConfig DNSTransportConfig `json:"transportConfig,omitempty"`
+}
+```
+
+Because of the limitation in CoreDNS that you can only have one `forward` 
+block per server and the TLS configuration for a forward block is global,
+DNS-over-TLS will be attempted for all upstreams configured for a zone by 
+default. Failure to establish a TLS connection for an upstream will result in a 
+log record indicating that a TLS connection could not be established.
+
+A cluster administrator can set only `Transport` and `ServerName` to allow
+the system certificates to be used for upstream resolver certificate
+validation. They can additionally set `CABundle` to use their own CA (or
+set of CAs) to validate upstream resolver certificates.
+
+### User Stories
+
+#### As a cluster administrator, I need to configure TLS for forwarded DNS requests so that security and integrity can be maintained
+
+To satisfy this use case for the default zone, the cluster administrator can 
+set:
+* `spec.upstreamResolvers.transport`
+* `spec.upstreamResolvers.caBundle`
+* `spec.upstreamResolvers.servername`
+
+Example:
+```yaml
+spec:
+  upstreamResolvers:
+    transport: tls
+    tls:
+      caBundle: dns-tls-configmap
+      servername: dns.bar.com
+    upstreams:
+    - type: Network
+      address: 1.1.1.1
```

Comment:
Is it possible for the cluster administrator to specify TLS while using the upstreams from `/etc/resolv.conf`?  If so, does doing so have well defined semantics?  
```yaml
spec:
  upstreamResolvers:
    transport: tls
    tls:
      caBundle: dns-tls-configmap
      servername: dns.bar.com
    upstreams:
    - type: SystemResolvConf
```
Related discussion: https://github.com/openshift/enhancements/pull/987#discussion_r792184198

Reply:
@Miciah this has been fixed in the latest implementation. Any additional changes needed with regards to this comment?

Reply:
Are you saying the API does allow the user to specify an invalid configuration?  Is it feasible to revise the API not to allow that?  Alternatively, we could document this in the godoc and [add an admission check](https://github.com/openshift/kubernetes/blob/master/openshift-kube-apiserver/admission/customresourcevalidation/dns/validate_dns.go) to reject the invalid configuration.

Reply:
> Are you saying the API does allow the user to specify an invalid configuration?
Yes, `UpstreamResolvers` and `ForwardPlugin` use the same union type for defining TLS configuration so the API allows this configuration. As far as I know, there isn't a way to refine the API to not allow this because `UpstreamResolvers` allows individual upstreams to use `/etc/resolv.conf` or an IP:port combination. This block was added to throw an error if TLS is configured for `UpstreamResolvers` when one of the `Upstream` objects is configured with `SystemResolvConf`: https://github.com/openshift/cluster-dns-operator/pull/314/files#diff-167ddcf106346ad9d8ce57a77bc468db4c4f41eeace963fb8cf01952f7e18adbR223-R233

Is the method we're using to address this insufficient? If so, is the next step to add an admission check to validate the configuration ahead of time?

Reply:
It's preferable if the API doesn't allow invalid configuration to be expressed in the first place.  The next best thing is an admission check to reject invalid configuration at admission time (per my previous comment).  If the validation cannot be done in the API and absolutely must be done on admitted objects, then we can do that (and we do that with the ingress operator), but it makes for an unfortunate UX.  Significantly changing the API at this point may not be an option, so I suggest we add an admission check.  The good news is that we already have the admission plugin for `dnses.operator.openshift.io`, so we can add the additional admission check into that plugin.

Reply:
Admission check added here: https://github.com/openshift/kubernetes/pull/1247